nautobot/dcim/forms.py

```diff
@@ -2359,6 +2366,59 @@ class DeviceFilterForm(
     tags = TagFilterField(model)
 
 
+class DeviceAddToClustersForm(BootstrapMixin, forms.Form):
+    """Form for adding a device to one or more clusters."""
+
+    location = DynamicModelChoiceField(
+        queryset=Location.objects.all(),
+        required=False,
+    )
+    cluster_type = DynamicModelChoiceField(
+        queryset=ClusterType.objects.all(),
+        required=False,
+    )
+    cluster_group = DynamicModelChoiceField(
+        queryset=ClusterGroup.objects.all(),
+        required=False,
+        query_params={"location": "$location"},
+    )
+    clusters = DynamicModelMultipleChoiceField(
+        queryset=Cluster.objects.all(),
+        required=True,
+        query_params={
+            "cluster_group": "$cluster_group",
+            "cluster_type": "$cluster_type",
+            "location": "$location",
+        },
+    )
+
+    def __init__(self, device, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        def disable_fields(help_text: str):
+            self.fields["location"].disabled = True
+            self.fields["cluster_type"].disabled = True
+            self.fields["cluster_group"].disabled = True
+            self.fields["clusters"].disabled = True
+            self.fields["clusters"].help_text = help_text
+
+        if self.fields["clusters"].queryset.exists():
+            available_clusters = Cluster.objects.exclude(pk__in=device.clusters.values_list("pk", flat=True)).order_by(
+                "name"
+            )
+            if not available_clusters.exists():
+                disable_fields("This device already belongs to all available clusters.")
```

Comment:
This is clever, though I'm curious how often it will actually come up in practice?